samples/tutorials/mnist/02_Local_Development_with_Docker_Image_Components.ipynb

```diff
@@ -0,0 +1,548 @@
+      "  warnings.warn(_CLOUD_SDK_CREDENTIALS_WARNING)\n"
+     ]
+    },
+    {
+     "data": {
+      "text/plain": [
+       "'gcr.io/kubeflow-pipeline-fantasy/mnist_training_kf_pipeline@sha256:3a3cf021de932b1104e639e42e69cb937fc56c8af8bd79fa0633bb2e44763f69'"
+      ]
+     },
+     "execution_count": 7,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "IMAGE_NAME=\"mnist_training_kf_pipeline\"\n",
+    "TAG=\"latest\" # \"v_$(date +%Y%m%d_%H%M%S)\"\n",
+    "\n",
+    "GCR_IMAGE=\"gcr.io/{PROJECT_ID}/{IMAGE_NAME}:{TAG}\".format(\n",
+    "    PROJECT_ID=PROJECT_ID,\n",
+    "    IMAGE_NAME=IMAGE_NAME,\n",
+    "    TAG=TAG\n",
+    ")\n",
+    "\n",
+    "builder = kfp.containers._container_builder.ContainerBuilder(\n",
+    "    gcs_staging=GCS_BUCKET + \"/kfp_container_build_staging\")\n",
+    "\n",
+    "image_name = kfp.containers.build_image_from_working_dir(\n",
+    "    image_name=GCR_IMAGE,\n",
+    "    working_dir='./tmp/components/mnist_training/',\n",
+    "    builder=builder\n",
+    ")\n",
+    "\n",
+    "image_name"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### If you want to use docker to build the image\n",
+    "Run the following in a cell\n",
+    "```bash\n",
+    "%%bash -s \"{PROJECT_ID}\"\n",
+    "\n",
+    "IMAGE_NAME=\"mnist_training_kf_pipeline\"\n",
+    "TAG=\"latest\" # \"v_$(date +%Y%m%d_%H%M%S)\"\n",
+    "\n",
+    "# Create script to build docker image and push it.\n",
+    "cat > ./tmp/components/mnist_training/build_image.sh <<HERE\n",
+    "PROJECT_ID=\"${1}\"\n",
+    "IMAGE_NAME=\"${IMAGE_NAME}\"\n",
+    "TAG=\"${TAG}\"\n",
+    "GCR_IMAGE=\"gcr.io/\\${PROJECT_ID}/\\${IMAGE_NAME}:\\${TAG}\"\n",
+    "docker build -t \\${IMAGE_NAME} .\n",
+    "docker tag \\${IMAGE_NAME} \\${GCR_IMAGE}\n",
+    "docker push \\${GCR_IMAGE}\n",
+    "docker image rm \\${IMAGE_NAME}\n",
+    "docker image rm \\${GCR_IMAGE}\n",
+    "HERE\n",
+    "\n",
+    "cd tmp/components/mnist_training\n",
+    "bash build_image.sh\n",
+    "```"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Define each component\n",
+    "Define a component by creating an instance of `kfp.dsl.ContainerOp` that describes the interactions with the Docker container image created in the previous step. You need to specify the component name, the image to use, the command to run after the container starts, the input arguments, and the file outputs. ."
```

Comment:
Remove extra period

Reply:
Is the command required (seems like something that the container may define by default)? Are the outputs guaranteed to be files?

Reply:
Since the no default entry point is defined in the Dockerfile above, I put it here. But sure, it can possibly be defined in the Dockerfile. And for the outputs, I think it need to be in the files (could be the raw content or the string path to the content).

Reply:
To make clearer, I have changed it to

-> 

### Define each component
Define a component by creating an instance of `kfp.dsl.ContainerOp` that describes the interactions with the Docker container image created in the previous step. You need to specify 
- component name
- the image to use
- the command to run after the container starts (If None, uses default CMD in defined in container.)
- the input arguments
- the file outputs (In the `app.py` above, the path of the trained model is written to `/output.txt`.)